File: pfi_sorter/turn_detection.py
"""Conservative, read-only GPS turn proposals.

The functions in this module never alter grouping decisions.  Coordinates are
used only for distance calculations and are deliberately absent from results.
"""

from __future__ import annotations

from collections import Counter
from datetime import datetime, timezone
from math import asin, cos, radians, sin, sqrt
from statistics import median
from typing import Any, Mapping, Sequence


DEFAULT_GPS_OPTIONS = {
    "gpsWindowSize": 3,
    "gpsMinDisplacementMeters": 4.0,
    "gpsMaxClusterRadiusMeters": 3.0,
    "gpsMinSignalRatio": 2.0,
    "gpsMaxGapSeconds": 30.0,
}


def _get(record: Any, key: str) -> Any:
    if isinstance(record, Mapping):
        return record.get(key)
    return getattr(record, key, None)


def _time_ms(value: Any) -> float | None:
    if value is None:
        return None
    if isinstance(value, datetime):
        if value.tzinfo is None:
            value = value.replace(tzinfo=timezone.utc)
        return value.timestamp() * 1000
    if isinstance(value, (int, float)):
        return float(value)
    return None


def haversine_meters(a: tuple[float, float], b: tuple[float, float]) -> float:
    """Return great-circle distance in metres."""
    lat1, lon1, lat2, lon2 = map(radians, (*a, *b))
    dlat, dlon = lat2 - lat1, lon2 - lon1
    value = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlon / 2) ** 2
    return 2 * 6_371_000 * asin(sqrt(value))


def _cluster(points: list[tuple[float, float]]) -> tuple[tuple[float, float], float]:
    centre = (median(point[0] for point in points), median(point[1] for point in points))
    return centre, max(haversine_meters(centre, point) for point in points)


def _runs(records: Sequence[Any], tolerance: float) -> list[dict[str, Any]]:
    runs: list[dict[str, Any]] = []
    current = None
    for index in range(1, len(records)):
        previous, item = records[index - 1], records[index]
        source = _get(item, "altitudeSource") or _get(item, "altitude_source")
        previous_source = _get(previous, "altitudeSource") or _get(previous, "altitude_source")
        a, b = _get(previous, "altitude"), _get(item, "altitude")
        direction = 0 if a is None or b is None or source not in {"relative", "absolute"} or source != previous_source else (1 if b - a > tolerance else -1 if a - b > tolerance else 0)
        if direction and current and current["direction"] == direction and current["source"] == source and current["end"] == index - 1:
            current["end"] = index
            current["steps"] += 1
        elif direction:
            current = {"start": index - 1, "end": index, "steps": 1, "direction": direction, "source": source}
            runs.append(current)
        else:
            current = None
    return runs


def analyze_gps_turns(records: Sequence[Any], settings: Mapping[str, Any] | None = None) -> dict[str, Any]:
    """Return proposals and aggregate rejection reasons for ordered records."""
    options = dict(DEFAULT_GPS_OPTIONS)
    options.update(settings or {})
    tolerance = float(options.get("altitudeTolerance", 0.75))
    min_steps = int(options.get("altitudeMinSteps", 2))
    min_span = float(options.get("altitudeMinSpan", 5.0))
    suppression = int(options.get("altitudeMarkerSuppression", 2))
    marker_pitch = float(options.get("markerPitch", -90))
    pitch_tolerance = float(options.get("tolerance", 2.0))
    reasons: Counter[str] = Counter()
    proposals: list[dict[str, Any]] = []
    runs = _runs(records, tolerance)
    sustained = [run for run in runs if run["steps"] >= min_steps and abs(_get(records[run["end"]], "altitude") - _get(records[run["start"]], "altitude")) >= min_span]
    if len(sustained) < 2:
        reasons["insufficient-altitude"] += 1
    for prior, nxt in zip(sustained, sustained[1:]):
        if prior["direction"] == nxt["direction"]:
            reasons["same-direction"] += 1
            continue
        if prior["source"] != nxt["source"] or prior["end"] >= nxt["start"]:
            reasons["inconsistent-altitude-source"] += 1
            continue
        boundary = nxt["start"]
        lo, hi = max(0, boundary - suppression), min(len(records), boundary + suppression + 1)
        if any((pitch := _get(record, "pitch")) is not None and abs(abs(pitch) - abs(marker_pitch)) <= pitch_tolerance for record in records[lo:hi]):
            reasons["nearby-pitched-down-marker"] += 1
            continue
        window = int(options["gpsWindowSize"])
        prior_indices = [i for i in range(prior["start"], prior["end"] + 1) if _get(records[i], "latitude") is not None and _get(records[i], "longitude") is not None][-window:]
        next_indices = [i for i in range(nxt["start"], nxt["end"] + 1) if _get(records[i], "latitude") is not None and _get(records[i], "longitude") is not None][:window]
        if len(prior_indices) < window or len(next_indices) < window:
            reasons["insufficient-gps"] += 1
            continue
        evidence_indices = list(range(prior["start"], nxt["end"] + 1))
        times = [_time_ms(_get(records[i], "captureTime") or _get(records[i], "capture_datetime") or _get(records[i], "captureDate")) for i in evidence_indices]
        if any(value is None for value in times):
            reasons["missing-time"] += 1
            continue
        gaps = [(times[i] - times[i - 1]) / 1000 for i in range(1, len(times))]
        if any(gap <= 0 for gap in gaps):
            reasons["non-increasing-time"] += 1
            continue
        maximum_gap = max(gaps, default=0.0)
        if maximum_gap > float(options["gpsMaxGapSeconds"]):
            reasons["excessive-time-gap"] += 1
            continue
        prior_points = [(_get(records[i], "latitude"), _get(records[i], "longitude")) for i in prior_indices]
        next_points = [(_get(records[i], "latitude"), _get(records[i], "longitude")) for i in next_indices]
        prior_centre, prior_radius = _cluster(prior_points)
        next_centre, next_radius = _cluster(next_points)
        if max(prior_radius, next_radius) > float(options["gpsMaxClusterRadiusMeters"]):
            reasons["noisy-gps-cluster"] += 1
            continue
        displacement = haversine_meters(prior_centre, next_centre)
        threshold = max(float(options["gpsMinDisplacementMeters"]), float(options["gpsMinSignalRatio"]) * max(prior_radius, next_radius, 0.5))
        if displacement < threshold:
            reasons["insufficient-displacement"] += 1
            continue
        filename = lambda i: str(_get(records[i], "fileName") or _get(records[i], "filename") or getattr(_get(records[i], "file"), "name", ""))
        detected = max(nxt["end"], next_indices[-1])
        proposals.append({
            "boundaryIndex": boundary, "detectedAtIndex": detected, "boundaryFile": filename(boundary),
            "reason": "gps-horizontal-turn", "status": "proposed", "evidence": {
                "priorDirection": "up" if prior["direction"] > 0 else "down", "nextDirection": "up" if nxt["direction"] > 0 else "down",
                "priorAltitudeSpanM": abs(_get(records[prior["end"]], "altitude") - _get(records[prior["start"]], "altitude")),
                "nextAltitudeSpanM": abs(_get(records[nxt["end"]], "altitude") - _get(records[nxt["start"]], "altitude")),
                "gpsDisplacementM": displacement, "priorClusterRadiusM": prior_radius, "nextClusterRadiusM": next_radius,
                "priorGpsSamples": len(prior_indices), "nextGpsSamples": len(next_indices), "maximumTimeGapSeconds": maximum_gap,
            }
        })
    return {"proposals": proposals, "reasonCounts": dict(reasons)}


def propose_gps_turns(records: Sequence[Any], settings: Mapping[str, Any] | None = None) -> list[dict[str, Any]]:
    """Return JSON-compatible GPS proposals without mutating ``records``."""
    return analyze_gps_turns(records, settings)["proposals"]

File: web-app/src/turnDetection.js
export const GPS_TURN_DEFAULTS = Object.freeze({
  gpsWindowSize: 3, gpsMinDisplacementMeters: 4, gpsMaxClusterRadiusMeters: 3,
  gpsMinSignalRatio: 2, gpsMaxGapSeconds: 30,
});

export function haversineMeters(a, b) {
  const toRad = (value) => value * Math.PI / 180;
  const [lat1, lon1, lat2, lon2] = [...a, ...b].map(toRad);
  const value = Math.sin((lat2 - lat1) / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin((lon2 - lon1) / 2) ** 2;
  return 2 * 6371000 * Math.asin(Math.sqrt(value));
}

const median = (values) => { const sorted = [...values].sort((a, b) => a - b); const middle = Math.floor(sorted.length / 2); return sorted.length % 2 ? sorted[middle] : (sorted[middle - 1] + sorted[middle]) / 2; };
function cluster(points) {
  const centre = [median(points.map((point) => point[0])), median(points.map((point) => point[1]))];
  return [centre, Math.max(...points.map((point) => haversineMeters(centre, point)))];
}
function altitudeRuns(records, tolerance) {
  const runs = []; let current = null;
  for (let index = 1; index < records.length; index += 1) {
    const previous = records[index - 1]; const item = records[index];
    const eligible = ['relative', 'absolute'].includes(item.altitudeSource) && item.altitudeSource === previous.altitudeSource && item.altitude != null && previous.altitude != null;
    const delta = eligible ? item.altitude - previous.altitude : 0;
    const direction = delta > tolerance ? 1 : delta < -tolerance ? -1 : 0;
    if (direction && current && current.direction === direction && current.source === item.altitudeSource && current.end === index - 1) { current.end = index; current.steps += 1; }
    else if (direction) { current = { start: index - 1, end: index, steps: 1, direction, source: item.altitudeSource }; runs.push(current); }
    else current = null;
  }
  return runs;
}
const fileName = (record) => record.fileName ?? record.filename ?? record.file?.name ?? '';
const timeMs = (record) => record.captureTime ?? record.captureDate?.getTime?.() ?? null;
const addReason = (counts, reason) => { counts[reason] = (counts[reason] ?? 0) + 1; };

export function analyzeGpsTurns(records, settings = {}) {
  // Every index returned by this function refers exclusively to the supplied
  // capture-time flight order. Callers must sort once before invoking it.
  const options = { ...GPS_TURN_DEFAULTS, ...settings }; const reasonCounts = {}; const proposals = [];
  const runs = altitudeRuns(records, options.altitudeTolerance ?? 0.75).filter((run) => run.steps >= (options.altitudeMinSteps ?? 2) && Math.abs(records[run.end].altitude - records[run.start].altitude) >= (options.altitudeMinSpan ?? 5));
  if (runs.length < 2) addReason(reasonCounts, 'insufficient-altitude');
  for (let pair = 1; pair < runs.length; pair += 1) {
    const prior = runs[pair - 1]; const next = runs[pair];
    if (prior.direction === next.direction) { addReason(reasonCounts, 'same-direction'); continue; }
    if (prior.source !== next.source || prior.end >= next.start) { addReason(reasonCounts, 'inconsistent-altitude-source'); continue; }
    const boundaryIndex = next.start; const suppression = options.altitudeMarkerSuppression ?? 2;
    const transitionStartIndex = prior.end; const transitionEndIndex = next.start;
    const suppressionStartIndex = Math.max(0, transitionStartIndex - suppression);
    const suppressionEndIndex = Math.min(records.length - 1, transitionEndIndex + suppression);
    const markerPitch = Math.abs(options.markerPitch ?? -90); const pitchTolerance = options.tolerance ?? 2;
    if (records.slice(suppressionStartIndex, suppressionEndIndex + 1).some((record) => record.pitch != null && Math.abs(Math.abs(record.pitch) - markerPitch) <= pitchTolerance)) { addReason(reasonCounts, 'nearby-pitched-down-marker'); continue; }
    const validGps = (index) => records[index].latitude != null && records[index].longitude != null;
    const priorIndices = Array.from({ length: prior.end - prior.start + 1 }, (_, offset) => prior.start + offset).filter(validGps).slice(-options.gpsWindowSize);
    const nextIndices = Array.from({ length: next.end - next.start + 1 }, (_, offset) => next.start + offset).filter(validGps).slice(0, options.gpsWindowSize);
    if (priorIndices.length < options.gpsWindowSize || nextIndices.length < options.gpsWindowSize) { addReason(reasonCounts, 'insufficient-gps'); continue; }
    const times = records.slice(prior.start, next.end + 1).map(timeMs);
    if (times.some((value) => value == null)) { addReason(reasonCounts, 'missing-time'); continue; }
    const gaps = times.slice(1).map((value, index) => (value - times[index]) / 1000);
    if (gaps.some((gap) => gap <= 0)) { addReason(reasonCounts, 'non-increasing-time'); continue; }
    const maximumTimeGapSeconds = Math.max(0, ...gaps);
    if (maximumTimeGapSeconds > options.gpsMaxGapSeconds) { addReason(reasonCounts, 'excessive-time-gap'); continue; }
    const points = (indices) => indices.map((index) => [records[index].latitude, records[index].longitude]);
    const [priorCentre, priorClusterRadiusM] = cluster(points(priorIndices)); const [nextCentre, nextClusterRadiusM] = cluster(points(nextIndices));
    if (Math.max(priorClusterRadiusM, nextClusterRadiusM) > options.gpsMaxClusterRadiusMeters) { addReason(reasonCounts, 'noisy-gps-cluster'); continue; }
    const gpsDisplacementM = haversineMeters(priorCentre, nextCentre);
    const noiseFloorM = Math.max(priorClusterRadiusM, nextClusterRadiusM, 0.5);
    const requiredDisplacementM = Math.max(options.gpsMinDisplacementMeters, options.gpsMinSignalRatio * noiseFloorM);
    const gpsSignalRatio = gpsDisplacementM / noiseFloorM;
    if (gpsDisplacementM < requiredDisplacementM) { addReason(reasonCounts, 'insufficient-displacement'); continue; }
    const detectedAtIndex = Math.max(next.end, nextIndices.at(-1));
    const evidenceStartIndex = prior.start; const evidenceEndIndex = next.end;
    proposals.push({ boundaryIndex, detectedAtIndex, boundaryFile: fileName(records[boundaryIndex]), detectedAtFile: fileName(records[detectedAtIndex]),
      evidenceStartIndex, evidenceEndIndex, evidenceStartFile: fileName(records[evidenceStartIndex]), evidenceEndFile: fileName(records[evidenceEndIndex]),
      transitionStartIndex, transitionEndIndex, transitionStartFile: fileName(records[transitionStartIndex]), transitionEndFile: fileName(records[transitionEndIndex]),
      suppressionStartIndex, suppressionEndIndex, requiredDisplacementM, gpsSignalRatio,
      reason: 'gps-horizontal-turn', status: 'proposed', evidence: {
      priorDirection: prior.direction > 0 ? 'up' : 'down', nextDirection: next.direction > 0 ? 'up' : 'down',
      priorAltitudeSpanM: Math.abs(records[prior.end].altitude - records[prior.start].altitude), nextAltitudeSpanM: Math.abs(records[next.end].altitude - records[next.start].altitude),
      gpsDisplacementM, priorClusterRadiusM, nextClusterRadiusM, priorGpsSamples: priorIndices.length, nextGpsSamples: nextIndices.length, maximumTimeGapSeconds,
    } });
  }
  return { proposals, reasonCounts };
}

export const proposeGpsTurns = (records, settings = {}) => analyzeGpsTurns(records, settings).proposals;
